### PR TITLE
Snackbar design

### PR DESCRIPTION
**Pull request by kolanse**
- Added a snackbar to notify users when an alarm is set 
- Please refer to issue  #352 for more information
 

WhatsApp Image 2021-07-22 at 22 37 09 (4)

WhatsApp Image 2021-07-22 at 22 37 09 (3)

WhatsApp Image 2021-07-22 at 22 37 09 (2)

WhatsApp Image 2021-07-22 at 22 37 09 (1)

WhatsApp Image 2021-07-22 at 22 37 09

### CODE REVIEW

File: app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt [hunk 1]
```diff
@@ -189,13 +203,52 @@ class AlarmsListActivity : AppCompatActivity() {
...
+        val rootView = window.decorView.rootView
+
+
+
+
+
+        store.sets().withLatestFrom<Boolean, Pair<Store.AlarmSet, Boolean>>(
+            store.uiVisible,
+            BiFunction { set, uiVisible -> set to uiVisible })
+            .subscribe { (set: Store.AlarmSet, uiVisible: Boolean) ->
```
Comment (yuriykulikov) [inline app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt:208]: Leaked disposable here. This subscription will not be disposed when the Fragment is destroyed

- [ ] Add disposable to the composite disposable of the fragment

File: app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt [hunk 2]
```diff
@@ -189,13 +203,52 @@ class AlarmsListActivity : AppCompatActivity() {
     override fun onStart() {
         super.onStart()
         configureTransactions()
+        configureSnackbar()
+    }
+
+    private fun configureSnackbar() {
+        val rootView = window.decorView.rootView
```
Comment (yuriykulikov) [inline app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt]: Can this fail? Does it make sense to move it inside the subscribe block?

File: app/src/main/java/com/better/alarm/presenter/ToastPresenter.kt [hunk 3]
```diff
@@ -1,105 +1,58 @@
...
+ * limitations under the License.
+ */
+
+package com.better.alarm.presenter
+
+import android.content.Context
+import android.widget.Toast
+import com.better.alarm.R
+import com.better.alarm.configuration.Store
+import com.better.alarm.util.formatToast
```
Comment (yuriykulikov) [inline app/src/main/java/com/better/alarm/presenter/ToastPresenter.kt:24]: These imports are not necessary here, right?

File: app/src/main/res/values-v21/themes_deep_blue.xml [hunk 4]
```diff
@@ -32,6 +36,16 @@
         <item name="drawerBackgroundColor">@color/deepblue_lighter_background</item>
         <item name="drawerClosedBackgroundColor">@color/deepblue_background</item>
     </style>
+    
+    <style name="Widget.App.SnackbarDefaultDeepBlueTheme" parent="" >
```
Comment (yuriykulikov) [inline app/src/main/res/values-v21/themes_deep_blue.xml]: Perhaps remove these empty lines?

File: app/src/main/java/com/better/alarm/presenter/ToastPresenter.kt [hunk 5]
```diff
@@ -1,105 +1,58 @@
...
+class ToastPresenter(private val store: Store, private val context: Context) {
+    private var sToast: Toast? = null
+
+    fun start() {
+        store.sets().withLatestFrom<Boolean, Pair<Store.AlarmSet, Boolean>>(store.uiVisible, BiFunction { set, uiVisible -> set to uiVisible })
+                .subscribe { (set: Store.AlarmSet, uiVisible: Boolean) ->
+                    if (!uiVisible && set.alarm.isEnabled) {
+                        popAlarmSetToast(context, set.millis)
+                    }
+
```
Comment (yuriykulikov) [inline app/src/main/java/com/better/alarm/presenter/ToastPresenter.kt:37]: Empty lines

**Review by yuriykulikov — CHANGES_REQUESTED** [review]
Looks good to me, but please prepare the commit for merging by reverting the unnecessary empty lines and changes in unrelated files. Here is how you can do that, or at least this is how I do that :)

1. in the git console run `git reset HEAD^`
2. In the Android Studio go to the Commit tab and open all files which are changed one by one and check what has been changed
3. Revert changes in unrelated places, such as inserting empty lines and so on

Also, it would be nice if there was a test, but I guess it would be quite hard to make it reliable :(

File: app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt (hunk 2 above)
Comment (kolanse) [reply app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt]: Yeah it does makes sense.

File: app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt (hunk 1 above)
Comment (kolanse) [reply app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt:208]: This code is in the activity

Comment (kolanse) [reply app/src/main/java/com/better/alarm/presenter/AlarmsListActivity.kt:208]: Done . i added it in the onStop of the activity

**Comment by kolanse** [comment]
@yuriykulikov  i have made the recommended changes and pushed. do i need to create a new PR request. Also for the test. do you have a test written for the toastpresenter class. perhaps it might serve as a guide for me

**Comment by yuriykulikov** [comment]
@kolanse as you can see now github cannot merge the PR automatically.

Could you please do the following?

`git remote add upstream git@github.com:yuriykulikov/AlarmClock.git`

`git fetch upstream`

`git rebase upstream/redesign -i`

-squash everything- 

`git push origin HEAD -f`

These commands will update your PR, which will allow me to merge this.

As for the test, no I do not have a test for the toast thing :) If you don't have prior knowledge about espresso tests it will be very time consuming to implement a test for the snackbar. Let's merge this one first.

**Comment by kolanse** [comment]
@yuriykulikov  Thank you very much. i will follow your instructions. as for the espresso test. i will need to time to work on it because it might be time-consuming. but as you said let us merge this first. I will work on it later given the chance

**Comment by kolanse** [comment]
@yuriykulikov  i have force pushed as instructed and done the other instructions. Thank you

**Comment by yuriykulikov** [comment]
Merged. @kolanse thank you for the contribution!

**Comment by kolanse** [comment]
@yuriykulikov  you are welcome

**Pull request merged**